benchmarks/flex/normal_dist.d

```diff
@@ -0,0 +1,326 @@
+/+ dub.json:
+{
+    "name": "bench_flex_normal",
+    "dependencies": {
+        "mir": {"path":"../.."},
+        "hap": "1.0.0-rc.2.1",
+        "dstats": "1.0.3",
+        "atmosphere": "0.1.7"
+    },
+    "dflags-ldc": ["-mcpu=native"]
+}
++/
+/+
+$ ldc-git --version
+LDC - the LLVM D compiler (798cda):
+  based on DMD v2.071.2-b1 and LLVM 3.8.1
+  built with DMD64 D Compiler v2.071.1
+  Default target: x86_64-unknown-linux-gnu
+  Host CPU: haswell
+  http://dlang.org - http://wiki.dlang.org/LDC
+
+$ dub run --build=release-nobounds --compiler=ldmd2-git --single benchmarks/flex/normal_dist.d
+
+boxMueller.naive   = 12 secs, 36 ms, and 998 μs
+boxMueller.hap     = 40 secs, 844 ms, 230 μs, and 6 hnsecs
+boxMueller.dstats  = 8 secs, 4 ms, 556 μs, and 4 hnsecs
+flexNormal.slow    = 14 secs, 54 ms, 592 μs, and 4 hnsecs
+flexNormal.medium  = 12 secs, 13 ms, 252 μs, and 9 hnsecs
+flexNormal.fast    = 11 secs, 174 ms, 573 μs, and 5 hnsecs
+ziggurat           = 3 secs, 538 ms, 75 μs, and 9 hnsecs
++/
+
+import mir.random.flex;
+
+/*
+$(WEB https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform, Box Muller-transform)
+References:
+    Butcher, J. C. "Random sampling from the normal distribution."
+    The Computer Journal 3.4 (1961): 251-253.
+    http://comjnl.oxfordjournals.org/content/3/4/251.full.pdf
+*/
+auto boxMueller(S, RNG)(S mu, S sigma, ref RNG gen)
+{
+    import mir.internal.math : log, cos, sin, sqrt;
+    import std.math : PI;
+    import std.random : uniform;
+
+    static S epsilon = S.min_normal;
+	static S two_pi = 2.0 * PI;
+
+	static S z0, z1;
+	static bool generate;
+	generate = !generate;
+
+	if (!generate)
+	   return z1 * sigma + mu;
+
+	S u1, u2;
+	do
+	{
+	   u1 = uniform(0.0L, 1.0L, gen);
+	   u2 = uniform(0.0L, 1.0L, gen);
+	}
+	while (u1 <= epsilon);
+
+	z0 = sqrt(-2.0 * log(u1)) * cos(two_pi * u2);
+	z1 = sqrt(-2.0 * log(u1)) * sin(two_pi * u2);
+	return z0 * sigma + mu;
+}
+
+auto genNormal(S)(S rho = 1.1)
+{
+    import std.math : exp, log, PI, sqrt;
+    S[] points = [-S.infinity, -1.5, 0, 1.5, S.infinity];
+    enum S halfLog2PI = S(0.5) * log(2 * PI);
+    auto f0 = (S x) => -(x * x) * S(0.5) - halfLog2PI;
+    auto f1 = (S x) => -x;
+    auto f2 = (S x) => S(-1);
+    return flex(f0, f1, f2, -0.5, points, rho);
+}
+
+__gshared float r = 0.0;
+
+void main()
+{
+    import std.datetime: benchmark, Duration;
+    import std.stdio : writefln;
+    import std.conv : to;
+
+    alias S = double;
+
+    auto flexNormalSlow = genNormal!S(1.3);
+    auto flexNormalMedium = genNormal!S(1.1);
+    auto flexNormalFast = genNormal!S(1.0001);
+
+    auto zigguratNormal = normal!(S, uint)();
+
+    // just pick any rng gen, it will have the same speed for all algorithms
+    import std.random : Mt19937;
+    auto gen = Mt19937(42);
+
+
+    auto bench = benchmark!(
+        { r += boxMueller!S(0, 1, gen); },
+        {
+            import hap.random.distribution : normal;
+            r += normal(0, 1, gen);
```

Comment:
From `hap.random` docs:

> Note that this function uses two variates from the uniform random number generator to generate a single normally-distributed variate. It is therefore an inefficient means of generating a large number of normally-distributed variates. If you wish to draw many variates from the normal distribution, it is better to use the range-based `normalDistribution` instead.

For fair speed comparison, you need to be creating a `NormalDistribution` range.  That should only be initialized once, i.e.

```D
auto hapNorm = normalDistribution(0, 1, gen);
```

... & then in your benchmark, use `hapNorm.front` and `hapNorm.popFront()` (you don't pass in the RNG on a per-call basis).